src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,929 @@
+      lo[i] == 0                (according to 2.4)
+
+      this leads to: r[x] == lo[x], for x <= i
+      while r[k] == 1 != lo[k] == 0, we can conclude that k > i
+
+      However, since:
+      lo[x] satisfies bits for 0 <= x < k:
+        According to 2.b.2.3, lo[x] == r[x] and r satisfies bits
+      zeros[k] == 0 (according to 2.b.2.1, r[k] == 1 and r satisfies bits)
+      lo[k]    == 0 (according to 2.b.2.2)
+
+      This contradicts the assumption that i is the largest bit index satisfying such conditions.
+
+    b.3. If j > i
+      ones[j] == v[j] (according to 2.7 since j > i)
+      v[j]    == 1    (according to 2.b.2)
+      r[j]    == 0    (according to 2.b.1)
+
+      This means that r[j] == 0 and ones[j] == 1, this contradicts the assumption that r
+      satisfies bits.
+
+    All cases lead to contradictions, which mean r < v is incorrect, which means that
+    r == v, which means the value v having the above form is the lowest value not smaller
+    than lo that satisfies bits.
+
+  3. Conclusion
+    Our objective now is to find the largest value i that satisfies:
+    - lo[x] satisfies bits for 0 <= x < i
+    - zeros[i] = 0
+    - lo[i] = 0
+
+    Call j the largest value such that lo[x] satisfies bits for 0 <= x < j. This
+    means that j is the smallest value such that lo[j] does not satisfy bits. We
+    call this the first violation. i then can be computed as the largest value
+    <= j such that:
+
+    zeros[i] == lo[i] == 0
+  */
+
+  // The algorithm depends on whether the first violation violates zeros or
+  // ones, if it violates zeros, we have the bit being 1 in zero_violation and
+  // 0 in one_violation. Since all higher bits are 0 in zero_violation and
+  // one_violation, we have zero_violation > one_violation. Similarly, if the
+  // first violation violates ones, we have zero_violation < one_violation.
+  if (zero_violation < one_violation) {
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 0 that should be a 1. Obviously, since the bit at that position in
+    // ones is 1, the same bit in zeros is 0. Which means this is the value of
+    // i we are looking for.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 1 0 0 1 0
+    //   zeros = 0 0 1 0 0 1 0 0
+    //    ones = 0 1 0 0 1 0 1 0
+    //   1-vio = 0 1 0 0 1 0 0 0
+    //   0-vio = 0 0 0 0 0 0 0 0
+    // Since the result must have the 2nd bit set, it must be at least:
+    //           1 1 0 0 0 0 0 0
+    // This value must satisfy zeros, because all bits before the 2nd bit have
+    // already satisfied zeros, and all bits after the 2nd bit are all 0 now.
+    // Just OR this value with ones to obtain the final result.
+
+    // first_violation is the position of the violation counting from the
+    // highest bit down (0-based), since i == 2, first_difference == 1
+    juint first_violation = count_leading_zeros(one_violation); // 1
```

Comment:
```suggestion
    juint first_violation = count_leading_zeros(one_violation);
```
What was the comment for?

Reply:
Ah, you want to say that `first_violation == 1`. Well you already say that above for the example... so not necessary I would say.